### ISSUE TITLE
ChannelBind and Linux eBPF

### ISSUE THREAD

**Issue by jonesmz** [coturn/coturn#759]
An idea for some other implementer to consider:

When a Channel is bound on the TURN allocation from a particular TURN Client to a particular remote destination, the subsequent packets that utilize the channel have characteristics that would allow for a Linux eBPF program attached to the network subsystem to lookup the first 4 bytes in a BPF map that has the information necessary to route the packet to the remote destination without the TURN server needing to be involved. The reverse case is also true, where a packet from the remote destination can have it's 5-tuple looked up in a map to identify the TURN Client and channel number.

This would allow for substantial improvements in latency and throughput for channel data packets that did not require translation to another protocol type (e.g. UDP<->UDP works fine, UDP<->TCP wouldn't)

**Comment by misi** [comment]
I fully agree that this will be beneficial in many things. I am happy to merge a PR with that.
But we should also consider that still the main use-case of the TURN server is the TCP 443 port, and there as you wrote TCP<->UDP it will not work this approach.

**Comment by jonesmz** [comment]
I know that for one coturn deployment at a large organization, TCP<->UDP is not enabled. But I certainly couldn't speak for other organizations.

That being said, it would not surprise me if somehow eBPF had some kind of support for the TCP<->UDP use case that I'm not aware of. My knowledge of eBPF comes from a very old kernel version, so there have probably been many improvements that I'm not aware of.

Either way, best of luck to whoever ends up implementing something of this nature.

**Comment by misi** [comment]
Anyway thanks for the proposal. Do you want to implement it? 
Unfortunately I am busy with other tasks so I don't think I can investigate it in the very close future..

**Comment by jonesmz** [comment]
Was this a bad idea or something?

I'm not going to implement it, but that doesn't seem like a justification for closing the ticket?

**Comment by misi** [comment]
Sorry I just felt you didn't answer to my last message and we have too many open issues. 
If you don't want to implement and nobody came to pick the idea then it will be anyhow archived and searchable even if it is closed. I said it is great idea, but we need someone who could implement it. I will leave it open for 6 month from now. Let's see what happens.

**Comment by stv0g** [comment]
I started to work on this for the client side. Eventually, I am also planning to integrate this into https://github.com/pion/turn and parts could be reused for coturn.

However, in the current state, Linux eBPF programs attached to the TC bpf filter do not support pushing/popping headers from a L4 UDP payload. I am currently trying to get some support from the BPF kernel mailing list to solve this issue:

https://lore.kernel.org/bpf/0150AA03-5A27-4AF4-8E59-A8AD7494CA0E@steffenvogel.de/T/#u

However, I am still waiting for any pointers and hints...
